src/main/scala/EssentVcd.scala

```diff
@@ -0,0 +1,286 @@
+    val registerDecs = registers flatMap {d: DefRegister => {
+      val regname = d.name
+      val regoldname = s"${regname}_old"
+      Seq(s"$regoldname = $regname;")
+    }}
+    val memDecs = memories map {m: DefMemory => {
+      s"${m.name}_old[${m.depth}] = ${m.name}[${m.depth}];"
+    }}
+    val modName = m.name
+    val ports_old = m.ports flatMap { p =>
+      //Removing signal names with "_" similar to verilator
+      if(p.name.contains("$_")) {
+        Seq()
+      }
+      else {
+        p.tpe match {
+          case ClockType => Seq(s"${p.name}_old = ${p.name};")
+          case _ =>
+            val ports_name = s"${p.name}_old"
+            Seq(s"$ports_name = ${p.name};")
+        }}}
+    if(modName == topName) {
+      writeLines(2, "//Assigning old values ")
+      writeLines(2, registerDecs)
+      writeLines(2, memDecs)
+      writeLines(2, ports_old)
+    }
+  }
+
+  //function for vcd multiple hierarchy
+  def hier_scope(allNamesAndTypes: Seq[(String, Type)],splitted: Seq[Seq[String]], indentlevel: Int, iden_code_hier: String) {
+
+    // This groups returns a Map( key -> Seq[Seq[String]]
+    val grouped = splitted groupBy {_.head }
+    grouped foreach { case (key, value) => {
+      val next_values = value map {_.tail}
+      val next_non_empty_values = next_values.filter(_.nonEmpty)
+      val uni_value = value.distinct
+      if(uni_value.size == 1) {
+        value foreach { v => if(v.size == 1) {
+          val iden_code = iden_code_hier + key
+          allNamesAndTypes map {
+            case (name , tpe ) =>
+              val new_name = name.replace(".","$")
+              if(new_name == iden_code) {
+                val width = bitWidth(tpe)
+                writeLines(indentlevel,s"""outfile << "$$var wire $width !${iden_code} $key $$end" << "\\n";""")
+              }}
+        }
+        }}
+      else {
+        writeLines(indentlevel, s""" outfile << "$$scope module $key $$end" << "\\n";""")
+        val iden_code_hier_new = iden_code_hier + key + "$"
+        hier_scope(allNamesAndTypes,next_non_empty_values,indentlevel+1,iden_code_hier_new)
+        writeLines(indentlevel,s""" outfile << "$$upscope $$end" << "\\n";""")
+      }
+    }
+    }
+  }
+
+  def declareoldvalues_all(circuit: Circuit): Unit = {
+    circuit.modules foreach {
+      case m: Module => declareoldvalues(m, topName)
+      case m: ExtModule => Seq()
+    }
+    allNamesAndTypes map {
+      case(name, tpe) =>
+        if(rn.nameToMeta(name).decType != ExtIO && rn.nameToMeta(name).decType != RegSet) {
+          val new_name = name.replace('.','$')
+          if(!new_name.contains("$_") && !new_name.contains("$next") && !new_name.startsWith("_")) {
+            writeLines(1, s"""${genCppType(tpe)} ${new_name}_old;""")
+          }
+        }
+    }
+  }
+
+  def initialize_old_values(circuit: Circuit): Unit = {
+    writeLines(1, "if(cycle_count == 0) {")
+    writeLines(3, s"""outfile << \"$$dumpvars\" << "\\n" ;""")
```

Comment:
If you use the triple quotes, the backslashes are not necessary

Reply:
I have updated with the suggestion in Vcd.Scala wherever possible. Thanks!